include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -0,0 +1,616 @@
+    char *start = detail::align_ptr_down(ptr, block_size_);
+    char *end   = start + block_size_;
+    return { start, end };
+  }
+
+  /**
+   * @brief Mark all memory in the given range as available; check partially covered blocks.
+   *
+   * All blocks completely covered by the given memory range are marked as available.
+   * The blocks which are only partially covered (at the beginning and at the end) are checked
+   * in the free tree and marked accordingly. The function assumes that the free tree has
+   * already been updated to reflect the availability.
+   */
+  void mark_as_available(void *ptr, size_t size) {
+    char *cptr = static_cast<char *>(ptr);
+
+    char *start = nullptr, *end = nullptr;
+    if (!is_block_aligned(cptr)) {
+      auto block = block_bounds(cptr);
+      start = free_mapped_.contains(block.first, block.second)
+              ? block.first : block.second;
+      // If the entire block is free then include it in the range - otherwsie point
+      // to the beginning of the next block.
+    } else {
+      start = cptr;
+    }
+    if (!is_block_aligned(cptr + size)) {
+      auto block = block_bounds(cptr + size);
+      end = free_mapped_.contains(block.first, block.second)
+              ? block.second : block.first;
+      // If the entire block is free then include it in this range - otherwise
+      // point to the end of the previous block.
+    } else {
+      end = cptr + size;
+    }
+    // It's possible that at this point end < start. It can happen if both ends of the range
+    // (ptr, ptr + size) lie within one block and that block is not avaialble - in this case
+    // `start` will point to the start of next block and `end` will point to end of the
+    // previous block, so they will become swapped. This is handled properly by
+    // set_block_availability.
+    set_block_avaialbility(start, end, true);
+  }
+
+  /**
+   * @brief Mark all blocks that overlap the given range as unavailable
+   */
+  void mark_as_unavailable(void *ptr, size_t size) {
+    char *cptr = static_cast<char *>(ptr);
+    char *start = detail::align_ptr_down(cptr, block_size_);
+    char *end = detail::align_ptr(cptr + size, block_size_);
+    set_block_avaialbility(start, end, false);
+  }
+
+
+  void set_block_avaialbility(char *start, char *end, bool availability) {
+    if (end <= start)  // NOTE it's possible that end < start - see mark_as_available
+      return;
+    va_region *va = va_find(start);
+    assert(va != nullptr);
+    assert(va == va_find(end-1));
+    va->set_block_availability(start, end, availability);
+  }
+
+  void map_storage(void *ptr, size_t size) {
+    CUdeviceptr dptr = reinterpret_cast<CUdeviceptr>(ptr);
+    auto *region = va_find(ptr);
+    assert(region);
+    CUdeviceptr region_start  = region->address_range.ptr();
+    CUdeviceptr region_end    = region->address_range.end();
+    ptrdiff_t offset = dptr - region_start;
+    int start_block_idx = offset / block_size_;
```

Comment:
Do we have a guarantee that the `dptr` is aligned to the block size?

Reply:
No - but if it's not, we round down to get to the beginning of the block.